distribution/xfer/download.go

```diff
@@ -81,6 +97,104 @@ type DownloadDescriptorWithRegistered interface {
 	Registered(diffID layer.DiffID)
 }
 
+// CollectGarbage removes any cached layers associated with downloads that are
+// still in progress
+func (ldm *LayerDownloadManager) CollectGarbage() error {
+	// we just flush the whole resumption cache and then save it
+	ldm.cachedLayerRefsLock.Lock()
+	for _, l := range ldm.cachedLayerRefs {
+		layer.ReleaseAndLog(ldm.layerStore, l)
+	}
+	ldm.cachedLayerRefs = make(map[string]layer.Layer)
+	ldm.cachedLayerRefsLock.Unlock()
+	ldm.saveLayerRefs()
+
+	return ldm.tm.CollectGarbage()
+}
+
+func (ldm *LayerDownloadManager) loadLayerRefs() {
+	data, err := ioutil.ReadFile(ldm.imagePullCheckpointsPath)
+	if err != nil {
+		logrus.Warnf("failed to read image pull checkpoint: %s", err)
+	}
+	checkpoint := map[string]string{}
+	err = json.Unmarshal(data, &checkpoint)
+	if err != nil {
+		logrus.Warnf("failed to unmarshal pull checkpoint: %s", err)
+	}
+	ldm.cachedLayerRefsLock.Lock()
+	defer ldm.cachedLayerRefsLock.Unlock()
+	for key, chainID := range checkpoint {
+		layer, err := ldm.layerStore.Get(layer.ChainID(chainID))
```

Comment:
I guess this mechanism of keeping references to the parent layers of partially downloaded layers is okay, since clearing the cache will release those references and allow the registered layers to be garbage collected. But I wonder if it would keep things simpler to instead store download artifacts for all layers in the pull until the entire pull succeeds, and reregister layers if necessary when resuming.